### PR TITLE
add SVM sensitivity example

### PR DESCRIPTION
**Pull request by be-apt**
As @matbesancon suggested, a practical example. Added notebook explains sensitivity analysis of data points in an SVM, 
 (borrowed from http://github.com/matbesancon/SimpleSVMs.jl).

Although in some examples I found points far away from hyperplane highly sensitive, in most cases points it followed the norm: points closest to the separating hyperplane are most sensitive.

Next thing I'll do a sensitvity analysis of several hyperplanes

### CODE REVIEW

**Comment by matbesancon** [comment]
1. Could you extract a .jl file from the notebook, they are easier to manage in git.
2. You are using dependencies that don't seem to be in the Project, maybe add a Project.toml in the example repo?

**Comment by AKS1996** [comment]
added the julia file

**Comment by matbesancon** [comment]
Don't you get an error at the cell "begin differentiating"?
image

**Comment by AKS1996** [comment]
forgot to mention, this works on https://github.com/jump-dev/DiffOpt.jl/tree/use_matrix_opt_interface branch and this repo https://github.com/AKS1996/MatrixOptInterface.jl
too many fixes to be added to master :D

**Comment by matbesancon** [comment]
Ok, so maybe we should finish merging #37 first :)

**Comment by matbesancon** [comment]
@AKS1996 I think we can finish this one now

**Comment by matbesancon** [comment]
we should probably remove the notebook altogether

**Comment by AKS1996** [comment]
worked on the example a bit. I'm wasnt able to comprehend sensitivity with regularization. 
Without it, differentiating it much easy

**Comment by AKS1996** [comment]
by the way, this PR seems very old :smile: 
too many changes since it started

**Comment by matbesancon** [comment]
@AKS1996 this still errors out (`backward_conic!` undefined)

**Comment by matbesancon** [comment]
The examples should be run in the documentation phase, we don't want to have them error out because we forgot them

**Review by matbesancon — CHANGES_REQUESTED** [review]
add to docs and adapt to current API

**Comment by AKS1996** [comment]
Good to go now?

**Comment by matbesancon** [comment]
@AKS1996 I am looking into running the example directly in Documenter, that way we make sure is actually run every time instead of manually making sure of it

**Comment by matbesancon** [comment]
I went through the example. I think what happens with the differentiation needs better explanation. What do we mean by "changing the labels" and "changing both the labels and predictors"? Change them or compute the gradient with respect to them.
More explanations are required on what happens in the last parts, because this is where DiffOpt is used

**Comment by matbesancon** [comment]
github action is still broken. We should really register MatrixOptInterface it's a pain to deal with unregistered deps

**Comment by matbesancon** [comment]
@AKS1996 I unbroke my broken GH action :sweat_smile: so you can look at my comment above and change the last part of the tutorial

**Comment by matbesancon** [comment]
bump :)

**Comment by AKS1996** [comment]
I've update the docs. Didn't change the examples since it contained mostly code.

File: docs/src/sensitivity-analysis-svm.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+![svg](svm_separating.svg)
+
+# Experiments
+Now that we've solved the SVM, we can compute the sensitivity of program variables -- the separating hyperplane in our case -- with respect to perturbations in problem data -- the data points -- using DiffOpt. For illustration, we've explored two questions:
+
+- How does a change in labels of the data points (`y=1` to `y=-1`, and vice versa) affects the position of the hyperplane? This is achieved by finding gradient of `w`, `b` with respect to `y[i]`, classification label of the ith data point (do not confuse `y[i]` with the array of dual variables.)
+- How does a change in coordinates of the data points, `X`, affects the position of the hyperplane? This is achieved by finding gradient of `w`, `b` with respect to `X[i]`, 2D coordinates of the data points.
+
+Note that the SVM program can be modelled as a conic optimization program
```
Comment (matbesancon) [inline docs/src/sensitivity-analysis-svm.md:126]:
```suggestion
Note that the SVM program can be modelled as a conic optimization program:
```

File: docs/src/sensitivity-analysis-svm.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+```
+
+Add the constraints.
+```@example 1
+MOI.add_constraint(
+    model,
+    MOI.VectorAffineFunction(
+        MOI.VectorAffineTerm.(1:N, MOI.ScalarAffineTerm.(1.0, l)), zeros(N)
+    ), 
+    MOI.Nonnegatives(N)
```
Comment (matbesancon) [inline docs/src/sensitivity-analysis-svm.md:65]:
```suggestion
    MOI.VectorAffineFunction(
        MOI.VectorAffineTerm.(1:N, MOI.ScalarAffineTerm.(1.0, l)), zeros(N),
    ), 
    MOI.Nonnegatives(N),
```

File: docs/src/sensitivity-analysis-svm.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+LinearAlgebra.normalize!(∇)
+nothing # hide
+```
+
+Visualize point sensitivities with respect to separating hyperplane. Note that the gradients are normalized.
+```@example 1
+p2 = Plots.scatter(
+    X[:,1], X[:,2], 
+    color = [yi > 0 ? :red : :blue for yi in y], label = "",
+    markersize = ∇ * 20
```
Comment (matbesancon) [inline docs/src/sensitivity-analysis-svm.md:192]:
```suggestion
    markersize = ∇ * 20,
```

File: docs/src/sensitivity-analysis-svm.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+end
+LinearAlgebra.normalize!(∇)
+```
+
+We can visualize point sensitvity with respect to the separating hyperplane. Note that the gradients are normalized.
+```@example 1
+p3 = Plots.scatter(
+    X[:,1], X[:,2], 
+    color = [yi > 0 ? :red : :blue for yi in y], label = "",
+    markersize = ∇ * 20
```
Comment (matbesancon) [inline docs/src/sensitivity-analysis-svm.md:232]:
```suggestion
    markersize = ∇ * 20,
```

File: docs/src/sensitivity-analysis-svm.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+Plots.yaxis!(p, (-2, 4.5))
+Plots.plot!(p, svm_x, svm_y, label = "loss = $(round(loss, digits=2))", width=3)
+Plots.savefig("svm_separating.svg")
+nothing # hide
+```
+
+![svg](svm_separating.svg)
+
+# Experiments
+Now that we've solved the SVM, we can compute the sensitivity of program variables -- the separating hyperplane in our case -- with respect to perturbations in problem data -- the data points -- using DiffOpt. For illustration, we've explored two questions:
```
Comment (matbesancon) [inline docs/src/sensitivity-analysis-svm.md:121]:
```suggestion
Now that we've solved the SVM, we can compute the sensitivity of optimal values -- the separating hyperplane in our case -- with respect to perturbations of the problem data -- the data points -- using DiffOpt. For illustration, we've explored two questions:
```

File: docs/src/sensitivity-analysis-svm.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+- How does a change in labels of the data points (`y=1` to `y=-1`, and vice versa) affects the position of the hyperplane? This is achieved by finding gradient of `w`, `b` with respect to `y[i]`, classification label of the ith data point (do not confuse `y[i]` with the array of dual variables.)
```
Comment (matbesancon) [inline docs/src/sensitivity-analysis-svm.md:123]:
```suggestion
- How does a change in labels of the data points (`y=1` to `y=-1`, and vice versa) affect the position of the hyperplane? This is achieved by finding the gradient of `w`, `b` with respect to `y[i]`, the classification label of the ith data point.
```

Comment (matbesancon) [reply docs/src/sensitivity-analysis-svm.md:123]: This one is tricky to justify with DiffOpt: there is no notion of an infinitesimal perturbation for the labels, I would be for removing it

File: docs/src/sensitivity-analysis-svm.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+- How does a change in coordinates of the data points, `X`, affects the position of the hyperplane? This is achieved by finding gradient of `w`, `b` with respect to `X[i]`, 2D coordinates of the data points.
+
+Note that the SVM program can be modelled as a conic optimization program:
```
Comment (matbesancon) [inline docs/src/sensitivity-analysis-svm.md:126]:
```suggestion
Note that finding the optimal SVM can be modelled as a conic optimization problem:
```

File: docs/src/sensitivity-analysis-svm.md (hunk 6 above)
Comment (AKS1996) [reply docs/src/sensitivity-analysis-svm.md:123]: yep. i didn't have any clue about perturbation in categorical data, but the change in value of the label (although not smooth) affects the hyperplane position

**Comment by joaquimg** [comment]
Maybe wait for #81 before merging this, since the interface is changing.

**Comment by matbesancon** [comment]
Yes agreed, no need to burden #81 with more changes to manage

**Comment by matbesancon** [comment]
@be-apt the PR needs to be adapted after the API changed introduced in #81

**Comment by be-apt** [comment]
Yep. Working on it

**Comment by be-apt** [comment]
dont know if changing the `db_cache` qualifies as mutating the model. removed the bangs again from forward and backward

**Comment by matbesancon** [comment]
> dont know if changing the db_cache qualifies as mutating the model

No, it's essentially an implementation detail

**Comment by matbesancon** [comment]
Congrats thanks @be-apt !
